src/app/core/services/rundown-timing.service.ts

```diff
@@ -0,0 +1,125 @@
+      ? segment.budgetDuration
+      : segment.parts.reduce((accumulatedPartDuration: number, part: Part) => accumulatedPartDuration + this.getPartDuration(part), 0)
+  }
+
+  public getRundownScheduleOffsetInMs(rundown: Rundown): number {
+    if (!rundown.isActive) {
+      const rundownPlannedEndEpochTime: number = this.getAggregatedEndEpochTime(rundown)
+      const estimatedRundownEndEpochTime: number = Date.now() + this.getAggregatedDurationInMs(rundown)
+      return estimatedRundownEndEpochTime - rundownPlannedEndEpochTime
+    }
+
+    const remainingDuration: number = this.getRemainingRundownDuration(rundown)
+    const estimatedEndEpochTime: number = Date.now() + remainingDuration
+    return estimatedEndEpochTime - this.getAggregatedEndEpochTime(rundown)
+  }
+
+  public getEndEpochTime(rundown: Rundown): number {
+    const expectedEndEpochTime: number | undefined = this.getExpectedEndEpochTime(rundown.timing)
+    if (expectedEndEpochTime) {
+      return expectedEndEpochTime
+    }
+
+    if (!rundown.isActive) {
+      const expectedDurationInMs: number = this.getExpectedDurationInMs(rundown.timing) ?? 0
+      return Date.now() + expectedDurationInMs
+    }
+
+    return Date.now() + this.getRemainingRundownDuration(rundown)
+  }
+
+  private getRemainingRundownDuration(rundown: Rundown): number {
+    const onAirSegment: Segment | undefined = rundown.segments.filter(segment => !segment.isUntimed).find(segment => segment.isOnAir)
+    const remainingOnAirSegmentBudgetDuration: number = onAirSegment?.budgetDuration ? Math.max(0, onAirSegment.budgetDuration - this.getOnAirSegmentPlayedDuration(onAirSegment)) : 0
+
+    const nextSegmentIndex: number = rundown.segments.findIndex(segment => segment.isNext)
+    if (nextSegmentIndex < 0) {
+      return remainingOnAirSegmentBudgetDuration
+    }
+    const segmentBudgetRemainingFromNext: number = rundown.segments
+      .slice(nextSegmentIndex)
+      .filter(segment => !segment.isOnAir && !segment.isUntimed)
+      .reduce((accumulatedEndEpochTime: number, segment: Segment) => accumulatedEndEpochTime + this.getSegmentDurationInMs(segment), 0)
+    return segmentBudgetRemainingFromNext + remainingOnAirSegmentBudgetDuration
+  }
+
+  private getOnAirSegmentPlayedDuration(onAirSegment: Segment): number {
+    const onAirPartIndex: number = onAirSegment.parts.findIndex(part => part.isOnAir)
+    const playedPartDuration: number = onAirSegment.parts
+      .slice(0, onAirPartIndex)
+      .reduce((accumulatedPartDuration: number, part: Part) => accumulatedPartDuration + this.partEntityService.getDuration(part), 0)
+    const onAirPartPlayedDuration: number = this.partEntityService.getPlayedDuration(onAirSegment.parts[onAirPartIndex])
+    return playedPartDuration + onAirPartPlayedDuration
+  }
+
+  private getPartDuration(part: Part): number {
+    if (part.isUntimed) {
+      return 0
+    }
+    return part.playedDuration > 0 ? part.playedDuration : part.expectedDuration ?? 0
```

Comment:
Be carefull when mixing more operations into ternaries. They quickly become obscure to read.
I think the use here is fine, but it's a slippery slope, hence the comment :)

Reply:
A good point :-)